garage/tf/models/base.py

```diff
@@ -210,20 +209,29 @@ def build(self, *inputs, name=None):
                     network = Network()
                     network._inputs = inputs
                     network._outputs = self._build(*inputs, name)
-        spec = self.network_output_spec()
-        if spec:
-            c = namedtuple(network_name,
-                           [*spec, 'input', 'output', 'inputs', 'outputs'])
+        in_spec = self.network_input_spec()
```

Comment:
this logic about `spec` is getting kind of complicated and i'm not sure i understand what's going on.

can you explain what the purpose of this code is? perhaps it would be good to break it into a well-documented helper function.

can you also help me understand the purpose of the input and output specs?

Reply:
Basically the `specs` tells us what the inputs and outputs are. There are 4 cases:
### Case 1: Single input/output model
Then we don't need any specs, since we don't need extra attribute for the networks. We only need `input,inputs,output,outputs` which are the default ones in network.
### Case 2: Single input/Multiple output model
We set `network_output_spec()` so that we can have assign "names" to the outputs, e.g. 
```python
def network_output_spec(self):
    return ['state', 'action']
```
Then we can do `model.networks['default'].state` and `model.networks['default'].action`
In this case, we don't need `network_input_spec()` since there is only one input. What we need is only `input`, which are already included by default.
### Case 3: Multiple input/Single output model
Same as above, but instead of having `network_output_spec`, we have `network_input_spec` this time.
### Case 4: Multiple input/output model
We have both `network_output_spec` and `network_input_spec`. Then we can have sth like
`model.networks['default'].state` and `model.networks['default'].action` as the outputs and `model.networks['default'].state_input` and `model.networks['default'].action_input` as the inputs, to make it more readable.

Reply:
this seems pretty complex -- is there a way we can always have an input/output spec, even for multi -input and -output models?

i think *always* having a spec would simplify the logic significantly.

Reply:
Do you mean we make `specs` not optional?
Right now by default the specs are a empty list, which means networks will only have `input,inputs,output,outputs`.

Reply:
i mean that you internally maintain the `specs` data structure regardless of whether the user has chosen to do multi input/output or not. then the single-input and single-output cases don't need to be handled with so much special logic.

the outer interface can stay the same.